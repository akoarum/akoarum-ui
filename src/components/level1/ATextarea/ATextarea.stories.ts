import { storiesOf } from '@storybook/vue'
import { action } from '@storybook/addon-actions'
import { withKnobs, text } from '@storybook/addon-knobs'
import { withInfo } from 'storybook-addon-vue-info'
import ATextarea from './ATextarea.vue'

storiesOf('level1/ATextarea', module)
  .addDecorator(withInfo)
  .addDecorator(withKnobs)
  .add('Normal', () => ({
    components: { ATextarea },
    props: {
      width: { default: text('幅', 'auto') },
      height: { default: text('高さ', '120px') }
    },
    data() {
      return {
        value: ''
      }
    },
    template: `
      <ATextarea v-model="value" :width="width" :height="height" @focus="focus" @blur="blur" />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur')
    }
  }), { info: {} })
  .add('Error', () => ({
    components: { ATextarea },
    props: {
      width: { default: text('幅', 'auto') },
      height: { default: text('高さ', '160px') }
    },
    data() {
      return {
        value: ''
      }
    },
    template: `
      <ATextarea
        v-model="value"
        :width="width"
        :height="height"
        aria-invalid="true"
        @focus="focus"
        @blur="blur"
      />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur')
    }
  }), { info: {} })
