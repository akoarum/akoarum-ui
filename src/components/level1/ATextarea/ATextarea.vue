<template>
  <textarea
    v-model="valueModel"
    :style="styled"
    class="aTextarea"
    @focus="handleFocus"
    @blur="handleBlur"
  />
</template>

<script lang="ts">
import { defineComponent, computed, SetupContext } from '@vue/composition-api'

interface Props {
  value: string
  width: string
  height: string
}

export default defineComponent({
  props: {
    value: { type: String, required: true },
    width: { type: String, default: 'auto' },
    height: { type: String, default: '160px' }
  },
  setup(props: Props, context: SetupContext) {
    const styled = computed<{ [key: string]: string }>(() => {
      return {
        '--width': props.width,
        '--height': props.height
      }
    })
    const valueModel = computed({
      get: (): string => props.value,
      set: (newValue: string) => {
        context.emit('input', newValue)
      }
    })
    const handleFocus = () => {
      context.emit('focus')
    }
    const handleBlur = () => {
      context.emit('blur')
    }

    return { styled, valueModel, handleFocus, handleBlur }
  }
})
</script>

<style lang="scss" scoped>
.aTextarea {
  --width: auto;
  --height: 160px;
  display: inline-block;
  box-sizing: border-box;
  border: 1px solid $color-border;
  box-shadow: none;
  padding: 8px 12px 9px;
  width: var(--width);
  height: var(--height);
  font-size: 16px;
  line-height: 1.3;
  outline: none;
  transition: border $duration $easing, background $duration $easing;

  &[aria-invalid='true'] {
    border-color: $color-caution;
    background: #fff7f7;
  }

  &:focus {
    border-color: $color-primary;
  }
}
</style>
