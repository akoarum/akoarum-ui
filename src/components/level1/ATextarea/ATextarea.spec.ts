import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueCompositionApi from '@vue/composition-api'
import ATextarea from './ATextarea.vue'

describe('ATextarea', () => {
  const localVue = createLocalVue()
  let wrapper

  localVue.use(VueCompositionApi)

  beforeEach(() => {
    wrapper = shallowMount(ATextarea, {
      propsData: {
        value: ''
      },
      localVue
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it('styled: propsで幅と高さをセットできる', () => {
      wrapper.setProps({ width: '100%', height: '140px' })
      expect(wrapper.vm.styled).toEqual({ '--width': '100%', '--height': '140px' })
    })
  })

  describe('event', () => {
    it('handleFocus: フォーカスイベントを通知する', () => {
      wrapper.vm.handleFocus()
      expect(wrapper.emitted().focus).toBeTruthy()
    })

    it('handleBlur: フォーカスが外れたことを通知する', () => {
      wrapper.vm.handleBlur()
      expect(wrapper.emitted().blur).toBeTruthy()
    })
  })
})
