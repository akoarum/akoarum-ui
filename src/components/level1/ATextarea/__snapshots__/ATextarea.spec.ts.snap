// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ATextarea initialized snapshot 1`] = `
<textarea
  class="aTextarea"
  style="--width: auto; --height: 160px;"
/>
`;
